cudatoolkit: prune broken symlinks in `postFixup`

ConnorBaker opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

As cudatoolkit is currently written, 11.8 introduces a broken symlink in `include` (also named `include`) and in `lib` (named `lib64`).

This trips up some consumers, (e.g., it causes the build of `tensorflow-gpu` to fail), and will be a problem when switching to 11.8 as the default.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Failures

- [x] TensorRT
    - Unrelated failure. Reproducible on `master` after adding `TensorRT-8.4.0.6.Linux.x86_64-gnu.cuda-11.6.cudnn8.3.tar.gz` to my store (per the guidance in the first error message, https://gist.github.com/ConnorBaker/9459ce1c741984959e78296c15e52f1e).
    - https://gist.github.com/ConnorBaker/882f8e425b585cb61f997259888b011b
- [x] tensorflow
    - Unrelated failure. Reproducible on `master`.
    - https://gist.github.com/ConnorBaker/06ceb965a933ae0659dfce58f9a8c654

ConnorBaker committed:
cudatoolkit: prune broken symlinks in `postFixup`

As cudatoolkit is currently written, 11.8 introduces a broken symlink in `include` (also named `include`) and in `lib` (named `lib64`).

This trips up some consumers, like `tensorflow-gpu`.

ConnorBaker commented:
Result of `nixpkgs-review pr 217322` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>caffeWithCuda</li>
    <li>cudaPackages.tensorrt</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.tensorrt</li>
    <li>xgboostWithCuda</li>
  </ul>
</details>
<details>
  <summary>30 packages built:</summary>
  <ul>
    <li>colmapWithCuda</li>
    <li>cudaPackages.cudatoolkit</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.nccl</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gpu-screen-recorder</li>
    <li>gpu-screen-recorder-gtk</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>hip-nvidia</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma (magma-cuda)</li>
    <li>nvtop</li>
    <li>nvtop-nvidia</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.pycuda</li>
    <li>python311Packages.pynvml</li>
    <li>python311Packages.pyrealsense2WithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

nviets reviewed (approved):
Approved - isn't this change for xgboost already included in #217333?

ConnorBaker commented:
> Approved - isn't this change for xgboost already included in #217333?

I couldn't figure out a way to test with `nixpkgs-review` besides rebasing on top of the branches containing the changes I needed. Specifying multiple PRs to test seemed to test them sequentially, so I'm stacking them.

If there's a better way to handle PRs which depend on each other, please let me know. My current plan is to continue to rebase this as the PRs it depends on are merged into master.

ConnorBaker commented:
Result of `nixpkgs-review pr 217322` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>35 packages built:</summary>
  <ul>
    <li>caffeWithCuda</li>
    <li>colmapWithCuda</li>
    <li>cudatoolkit (cudaPackages.cudatoolkit ,cudatoolkit_11)</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.nccl</li>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gpu-screen-recorder</li>
    <li>gpu-screen-recorder-gtk</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>hip-nvidia</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma (magma-cuda)</li>
    <li>nvtop</li>
    <li>nvtop-nvidia</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.pycuda</li>
    <li>python311Packages.pynvml</li>
    <li>python311Packages.pyrealsense2WithCuda</li>
    <li>xgboostWithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

ConnorBaker commented:
cc @NixOS/cuda-maintainers

samuela commented on pkgs/development/compilers/cudatoolkit/common.nix:
```diff
@@ -229,6 +229,12 @@ stdenv.mkDerivation rec {
     addOpenGLRunpath $out/cuda_sanitizer_api/compute-sanitizer/x86/*
     addOpenGLRunpath $out/target-linux-x64/*
+  '' + ''
+    # Prune broken symlinks which can cause problems with consumers of this package.
```
Can you please make this a Nix comment instead of a shell comment? Having these comments in the shell code means that they impact the derivation hash which means that it will cause rebuilds the next time we have to change the comment.

ConnorBaker commented:
Moved the comment out of the script and into Nix as requested.

Also removed the PR stack information from the OP and rebased on master instead since it can be merged independently of the others.

If you do run `nixpkgs-review`, you may see unrelated errors the PR stack included fixes for:
- https://github.com/NixOS/nixpkgs/pull/217330
- https://github.com/NixOS/nixpkgs/pull/217333

ConnorBaker commented:
@samuela would you mind taking a look if you have a chance?

samuela commented:
btw do TF/JAX build against CUDA 11.8 after this change? it would be great if we could finally upgrade our `cudaPackages` alias (separate, future PR)!

ConnorBaker commented:
I believe JAX does, though I remember Tensorflow failing at the very end of its very, very long build with an error about GLIBCXX. I haven't looked too much at it since it's been a constant failure across all of my PRs, but I wonder if it has something to do with compiler versions gone awry? Might need to make sure the derivation is being built `cudatoolkit.cc` and not whatever `stdenv` provides.

samuela commented:
ugh yeah TF is the worst to build... well that's not due to anything in this change. Seems like progress since IIRC before TF was failing at the beginning of the build with the same error that JAX had. So that's great news!

ConnorBaker commented:
Result of `nixpkgs-review pr 217322` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>caffeWithCuda</li>
    <li>colmapWithCuda</li>
    <li>cudatoolkit (cudaPackages.cudatoolkit ,cudatoolkit_11)</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.nccl</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gpu-screen-recorder</li>
    <li>gpu-screen-recorder-gtk</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>hip-nvidia</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma (magma-cuda)</li>
    <li>nvtop</li>
    <li>nvtop-nvidia</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.pycuda</li>
    <li>python311Packages.pynvml</li>
    <li>python311Packages.pyrealsense2WithCuda</li>
    <li>xgboostWithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

samuela commented:
Result of `nixpkgs-review pr 217322` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.caffeWithCuda</li>
    <li>truecrack-cuda</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>caffeWithCuda</li>
    <li>colmapWithCuda</li>
    <li>cudatoolkit (cudaPackages.cudatoolkit ,cudatoolkit_11)</li>
    <li>cudaPackages.cutensor</li>
    <li>cudaPackages.nccl</li>
    <li>forge</li>
    <li>gpu-burn</li>
    <li>gpu-screen-recorder</li>
    <li>gpu-screen-recorder-gtk</li>
    <li>gromacsCudaMpi</li>
    <li>gwe</li>
    <li>hip-nvidia</li>
    <li>katagoWithCuda</li>
    <li>librealsenseWithCuda</li>
    <li>magma (magma-cuda)</li>
    <li>nvtop</li>
    <li>nvtop-nvidia</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.pycuda</li>
    <li>python310Packages.pynvml</li>
    <li>python310Packages.pyrealsense2WithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.pycuda</li>
    <li>python311Packages.pynvml</li>
    <li>python311Packages.pyrealsense2WithCuda</li>
    <li>xgboostWithCuda</li>
    <li>xpraWithNvenc</li>
  </ul>
</details>

samuela commented:
LGTM thanks so much @ConnorBaker !

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-45/26397/1
